[pytorch][mobile] support for custom mobile build with dynamic dispatch

ljk53 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #34184 [pytorch][CI] add e2e mobile custom build jobs to CI
* **#34055 [pytorch][mobile] support for custom mobile build with dynamic dispatch**

Summary:
Enable custom mobile build with dynamic dispatch for OSS build.

It calls a python util script to calculate transitive dependencies from
the op dependency graph and the list of used root ops, then pass the
result as the op registration whitelist to aten codegen, so that only
these used ops are registered and kept at link time.

For custom build with dynamic dispatch to work correctly, it's critical
to have the accurate list of used ops. Current assumption is that only
those ops referenced by TorchScript model are used. It works well if
client code doesn't call libtorch API (e.g.  tensor methods) directly;
otherwise the extra used ops need to be added to the whitelist manually,
as shown by the HACK in prepare_model.py.

Also, if JIT starts calling extra ops independent of specific model,
then the extra ops need to be added to the whitelist as well.

Verified the correctness of the whole process with MobileNetV2:
```
TEST_CUSTOM_BUILD_DYNAMIC=1 test/mobile/custom_build/build.sh
```

Differential Revision: [D20193327](https://our.internmc.facebook.com/intern/diff/D20193327)

ljk53 committed:
[pytorch][mobile] support for custom mobile build with dynamic dispatch

Summary:
Enable custom mobile build with dynamic dispatch for OSS build.

It calls a python util script to calculate transitive dependencies from
the op dependency graph and the list of used root ops, then pass the
result as the op registration whitelist to aten codegen, so that only
these used ops are registered and kept at link time.

For custom build with dynamic dispatch to work correctly, it's critical
to have the accurate list of used ops. Current assumption is that only
those ops referenced by TorchScript model are used. It works well if
client code doesn't call libtorch API (e.g.  tensor methods) directly;
otherwise the extra used ops need to be added to the whitelist manually,
as shown by the HACK in prepare_model.py.

Also, if JIT starts calling extra ops independent of specific model,
then the extra ops need to be added to the whitelist as well.

Verified the correctness of the whole process with MobileNetV2:
```
TEST_CUSTOM_BUILD_DYNAMIC=1 test/mobile/custom_build/build.sh
```

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch][mobile] support for custom mobile build with dynamic dispatch"


Summary:
Enable custom mobile build with dynamic dispatch for OSS build.

It calls a python util script to calculate transitive dependencies from
the op dependency graph and the list of used root ops, then pass the
result as the op registration whitelist to aten codegen, so that only
these used ops are registered and kept at link time.

For custom build with dynamic dispatch to work correctly, it's critical
to have the accurate list of used ops. Current assumption is that only
those ops referenced by TorchScript model are used. It works well if
client code doesn't call libtorch API (e.g.  tensor methods) directly;
otherwise the extra used ops need to be added to the whitelist manually,
as shown by the HACK in prepare_model.py.

Also, if JIT starts calling extra ops independent of specific model,
then the extra ops need to be added to the whitelist as well.

Verified the correctness of the whole process with MobileNetV2:
```
TEST_CUSTOM_BUILD_DYNAMIC=1 test/mobile/custom_build/build.sh
```

[ghstack-poisoned]

iseeyuan commented:
"Also, if JIT starts calling extra ops independent of specific model,
then the extra ops need to be added to the whitelist as well." I just noticed there are some extra ops in the actual bytecode than from the get_opnames. Do you mean this? I could try some fixes on get_opnames.

bhosmer commented on test/mobile/custom_build/prepare_model.py:
```diff
@@ -13,5 +13,6 @@
 
 # Dump root ops used by the model (for custom build optimization).
 ops = torch.jit.export_opnames(traced_script_module)
+ops.append('aten::ones')  # HACK because predictor.cpp explicitly calls this!
```
Just a thought that if our eventual setup still needs to explicitly push roots into `ops` like this we'll probably want to institutionalize it, like have a global list called `EXTRA_ROOT_OPS` or whatever

ljk53 replied:
Yes, as I replied to iseeyuan, we need run static analysis against JIT code to find out common EXTRA_ROOT_OPS, if any.

However, this case is different - aten::ones is directly called from *client* code (the dummy predictor.cpp calls it to create all-one-tensor for testing purpose). Generally, user could write pure c++ client code to directly call functions or tensor methods, which won't be captured by extracting ops from model's bytecodes.

This is less a problem for Android, where we expect users to use the limited set of Java APIs. To actually solve this problem, we probably need ask users to run code analyzer against their client code to dump these extra root ops - that's why I feel there will be more support work to switch to dynamic dispatch - in static dispatch case these extra ops will be kept by linker automatically.

For CI purpose this one-off hack is probably fine? :)

bhosmer replied:
Question, does this mean every custom build is gonna get `aten::ones` whether they use it or not? Or just binaries that use the dummy `predictor.cpp`? (I'm guessing the former since this is a general script, but plz correct :) 

A single op isn't that big a deal I guess, but I think it's important to describe the situation more explicitly - something like your description above in a comment above a global that declares the op more explicitly. Text below is just your comment above, so may not be entirely correct, but this is the kind of thing I mean:

```
# We need to run static analysis against JIT code to find out common EXTRA_ROOT_OPS, if any.
# 
# However, this case is different - aten::ones is directly called from client code (the dummy
# predictor.cpp calls it to create all-one-tensor for testing purpose). Generally, user could
# write pure c++ client code to directly call functions or tensor methods, which won't be
# captured by extracting ops from model's bytecodes.
# 
# This is less a problem for Android, where we expect users to use the limited set of Java 
# APIs. To actually solve this problem, we probably need ask users to run code analyzer against 
# their client code to dump these extra root ops - that's why I feel there will be more support 
# work to switch to dynamic dispatch - in static dispatch case these extra ops will be kept by 
# linker automatically.
# 
# For CI purpose this one-off hack is probably fine? :)
EXTRA_CI_ROOT_OPS = ['aten::ones']

...
ops.extend(EXTRA_CI_ROOT_OPS)
```

ljk53 replied:
> Question, does this mean every custom build is gonna get `aten::ones` whether they use it or not? Or just binaries that use the dummy `predictor.cpp`? (I'm guessing the former since this is a general script, but plz correct :)

It's the latter - the script is under test/mobile/... which is for mobile CI specifically. That's why I don't feel so guilty to add this hack :)

ljk53 replied:
Added the comment to the test script to keep a record.

ljk53 committed:
Update on "[pytorch][mobile] support for custom mobile build with dynamic dispatch"


Summary:
Enable custom mobile build with dynamic dispatch for OSS build.

It calls a python util script to calculate transitive dependencies from
the op dependency graph and the list of used root ops, then pass the
result as the op registration whitelist to aten codegen, so that only
these used ops are registered and kept at link time.

For custom build with dynamic dispatch to work correctly, it's critical
to have the accurate list of used ops. Current assumption is that only
those ops referenced by TorchScript model are used. It works well if
client code doesn't call libtorch API (e.g.  tensor methods) directly;
otherwise the extra used ops need to be added to the whitelist manually,
as shown by the HACK in prepare_model.py.

Also, if JIT starts calling extra ops independent of specific model,
then the extra ops need to be added to the whitelist as well.

Verified the correctness of the whole process with MobileNetV2:
```
TEST_CUSTOM_BUILD_DYNAMIC=1 test/mobile/custom_build/build.sh
```

Differential Revision: [D20193327](https://our.internmc.facebook.com/intern/diff/D20193327)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch][mobile] support for custom mobile build with dynamic dispatch"


Summary:
Enable custom mobile build with dynamic dispatch for OSS build.

It calls a python util script to calculate transitive dependencies from
the op dependency graph and the list of used root ops, then pass the
result as the op registration whitelist to aten codegen, so that only
these used ops are registered and kept at link time.

For custom build with dynamic dispatch to work correctly, it's critical
to have the accurate list of used ops. Current assumption is that only
those ops referenced by TorchScript model are used. It works well if
client code doesn't call libtorch API (e.g.  tensor methods) directly;
otherwise the extra used ops need to be added to the whitelist manually,
as shown by the HACK in prepare_model.py.

Also, if JIT starts calling extra ops independent of specific model,
then the extra ops need to be added to the whitelist as well.

Verified the correctness of the whole process with MobileNetV2:
```
TEST_CUSTOM_BUILD_DYNAMIC=1 test/mobile/custom_build/build.sh
```

Differential Revision: [D20193327](https://our.internmc.facebook.com/intern/diff/D20193327)

[ghstack-poisoned]

ljk53 commented:
> "Also, if JIT starts calling extra ops independent of specific model,
> then the extra ops need to be added to the whitelist as well." I just noticed there are some extra ops in the actual bytecode than from the get_opnames. Do you mean this? I could try some fixes on get_opnames.

The problem you mentioned should be fixed (regardless of static dispatch custom build or dynamic dispatch custom build) :)

What I meant was potential case where model's bytecode doesn't fully cover all ops called from JIT, e.g. what if during model loading time some JIT code decides to call some ops directly? I can imagine some common preprocessing pass might do this.

One way to address this problem is to run static analysis against JIT codebase and find out. It will require a little work.

iseeyuan commented:
> > "Also, if JIT starts calling extra ops independent of specific model,
> > then the extra ops need to be added to the whitelist as well." I just noticed there are some extra ops in the actual bytecode than from the get_opnames. Do you mean this? I could try some fixes on get_opnames.
> 
> The problem you mentioned should be fixed (regardless of static dispatch custom build or dynamic dispatch custom build) :)
> 
> What I meant was potential case where model's bytecode doesn't fully cover all ops called from JIT, e.g. what if during model loading time some JIT code decides to call some ops directly? I can imagine some common preprocessing pass might do this.
> 
> One way to address this problem is to run static analysis against JIT codebase and find out. It will require a little work.

Are load-time operations embedded in __setstate__, which can also be collected in bytecode level. I'm not aware of other run-time op calls.

bhosmer reviewed (approved):
One inline thing about the commenting `aten::ones` dependency but LGTM :)

ljk53 committed:
Update on "[pytorch][mobile] support for custom mobile build with dynamic dispatch"


Summary:
Enable custom mobile build with dynamic dispatch for OSS build.

It calls a python util script to calculate transitive dependencies from
the op dependency graph and the list of used root ops, then pass the
result as the op registration whitelist to aten codegen, so that only
these used ops are registered and kept at link time.

For custom build with dynamic dispatch to work correctly, it's critical
to have the accurate list of used ops. Current assumption is that only
those ops referenced by TorchScript model are used. It works well if
client code doesn't call libtorch API (e.g.  tensor methods) directly;
otherwise the extra used ops need to be added to the whitelist manually,
as shown by the HACK in prepare_model.py.

Also, if JIT starts calling extra ops independent of specific model,
then the extra ops need to be added to the whitelist as well.

Verified the correctness of the whole process with MobileNetV2:
```
TEST_CUSTOM_BUILD_DYNAMIC=1 test/mobile/custom_build/build.sh
```

Differential Revision: [D20193327](https://our.internmc.facebook.com/intern/diff/D20193327)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch][mobile] support for custom mobile build with dynamic dispatch"


Summary:
Enable custom mobile build with dynamic dispatch for OSS build.

It calls a python util script to calculate transitive dependencies from
the op dependency graph and the list of used root ops, then pass the
result as the op registration whitelist to aten codegen, so that only
these used ops are registered and kept at link time.

For custom build with dynamic dispatch to work correctly, it's critical
to have the accurate list of used ops. Current assumption is that only
those ops referenced by TorchScript model are used. It works well if
client code doesn't call libtorch API (e.g.  tensor methods) directly;
otherwise the extra used ops need to be added to the whitelist manually,
as shown by the HACK in prepare_model.py.

Also, if JIT starts calling extra ops independent of specific model,
then the extra ops need to be added to the whitelist as well.

Verified the correctness of the whole process with MobileNetV2:
```
TEST_CUSTOM_BUILD_DYNAMIC=1 test/mobile/custom_build/build.sh
```

Differential Revision: [D20193327](https://our.internmc.facebook.com/intern/diff/D20193327)

[ghstack-poisoned]

mrshenli commented:
Hey @ljk53, looks like either this PR or the one below it has broken master. Do we have any fix coming for it?

https://app.circleci.com/jobs/github/pytorch/pytorch/4683824
https://app.circleci.com/jobs/github/pytorch/pytorch/4683798

ljk53 commented:
@mrshenli - looking - this might affect build but shouldn't break test.

ljk53 commented:
@mrshenli created #34188 and #34189 to revert #32814 and this PR, respectively: none of them fixed the failed CI :(

mrshenli commented:
@ljk53 Thanks a lot for investigating. Looks like that test is flaky. It also fails on #33860 but succeeded in several subsequent runs. Let me disable that.

ljk53 commented:
> > > "Also, if JIT starts calling extra ops independent of specific model,
> > > then the extra ops need to be added to the whitelist as well." I just noticed there are some extra ops in the actual bytecode than from the get_opnames. Do you mean this? I could try some fixes on get_opnames.
> > 
> > 
> > The problem you mentioned should be fixed (regardless of static dispatch custom build or dynamic dispatch custom build) :)
> > What I meant was potential case where model's bytecode doesn't fully cover all ops called from JIT, e.g. what if during model loading time some JIT code decides to call some ops directly? I can imagine some common preprocessing pass might do this.
> > One way to address this problem is to run static analysis against JIT codebase and find out. It will require a little work.
> 
> Are load-time operations embedded in `__setstate__`, which can also be collected in bytecode level? I'm not aware of other run-time op calls.

@iseeyuan Here is one example of calling ops during model loading time: https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/serialization/unpickler.cpp#L425
```
PickleOpCode Unpickler::readInstruction() {
  ...
  case PickleOpCode::BINPERSID: {
    ..
    tensor = at::empty({0}, options).set_(storage);
```
It calls `aten::empty()` and `aten::set_()`.

Today it works because these ops are pretty common and are also indirectly called by regular models - but there is no guarantee it will keep this way. For example, #34641 breaks dynamic dispatch CI in a subtle way. We probably need run code analyzer against torch/csrc to find out these common ops used by JIT.
